agera/src/main/java/com/google/android/agera/FunctionCompiler.java

```diff
@@ -32,16 +32,46 @@
 
 @SuppressWarnings({"unchecked, rawtypes"})
 final class FunctionCompiler implements FList, FItem {
+  private static final ThreadLocal<FunctionCompiler> compilers = new ThreadLocal<>();
+
+  @NonNull
+  static FunctionCompiler functionCompiler() {
+    FunctionCompiler compiler = compilers.get();
+    if (compiler == null) {
+      compiler = new FunctionCompiler();
+    } else {
+      // Remove compiler from the ThreadLocal to prevent reuse in the middle of a compilation.
+      // recycle(), called by compile(), will return the compiler here. ThreadLocal.set(null) keeps
+      // the entry (with a null value) whereas remove() removes the entry; because we expect the
+      // return of the compiler, don't use the heavier remove().
+      compilers.set(null);
+    }
+    return compiler;
+  }
+
+  @NonNull
+  private static void recycle(@NonNull final FunctionCompiler compiler) {
+    compiler.directives.clear();
+    compilers.set(compiler);
+  }
+
+  private static final int APPLY = 0;
```

Comment:
'Constants' cannot be Integers. Integers won't be constants in Java. The only reference type that can be a Java constant is String. `public static final Integer` fields take up dex counts and uses field read operations at run time.

ints are constants. The int range in use by this PR are all cached (at least in the Android implementation), so valueOf() [the boxing call] is quick and has zero allocations. Granted it's a range check + an array index lookup, which is slower than a field read, but the benefit of using a `public static final Integer` is still too small.